pkg/networkservice/common/timeout/server.go

```diff
@@ -21,80 +21,145 @@ import (
-			t.connections[connID] = ct
+
+	executor, _ := t.executors.LoadOrStore(connID, &serialize.Executor{})
+	<-executor.AsyncExec(func() {
+		// executor was possibly removed by `t.close()` at this moment, we need to store it back
+		exec, _ := t.executors.LoadOrStore(connID, executor)
+		if exec != executor {
+			err = errors.Errorf("race condition, parallel request execution: %v", connID)
+			return
+		}
+
+		if timer, ok := t.connections.Load(connID); ok {
+			if !timer.timer.Stop() {
+				logEntry.Warnf("connection has been timed out, re requesting: %v", connID)
+			}
+			close(timer.stopCh)
+			t.connections.Delete(connID)
 		}
-	})
-	return next.Server(ctx).Request(ctx, request)
-}
 
-func (t *timeoutServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
-	t.executor.AsyncExec(func() {
-		if timer, ok := t.connections[conn.GetId()]; ok {
-			timer.Stop()
-			delete(t.connections, conn.GetId())
+		conn, err = next.Server(ctx).Request(ctx, request)
+		if err != nil {
+			return
 		}
+
+		var timer *timer
+		timer, err = t.createTimer(ctx, conn)
+		if err != nil {
+			if _, closeErr := next.Server(ctx).Close(ctx, conn); closeErr != nil {
+				err = errors.Wrapf(err, "error attempting to close failed connection %v: %+v", connID, closeErr)
+			}
+			return
+		}
+
+		t.connections.Store(connID, timer)
 	})
-	return next.Server(ctx).Close(ctx, conn)
+
+	return conn, err
 }
 
-func (t *timeoutServer) createTimer(ctx context.Context, request *networkservice.NetworkServiceRequest) (*time.Timer, error) {
-	request = request.Clone()
-	expireTime, err := ptypes.Timestamp(request.GetConnection().GetPath().GetPathSegments()[request.GetConnection().GetPath().GetIndex()-1].GetExpires())
+func (t *timeoutServer) createTimer(ctx context.Context, conn *networkservice.Connection) (*timer, error) {
+	logEntry := log.Entry(ctx).WithField("timeoutServer", "createTimer")
+
+	expireTime, err := ptypes.Timestamp(conn.GetPath().GetPathSegments()[conn.GetPath().GetIndex()-1].GetExpires())
 	if err != nil {
 		return nil, err
 	}
-	duration := time.Until(expireTime)
-	conn := request.GetConnection().Clone()
-	return time.AfterFunc(duration, func() {
-		newCtx := extend.WithValuesFromContext(context.Background(), ctx)
-		if _, err := (*t.onTimeout).Close(newCtx, conn); err != nil {
-			trace.Log(newCtx).Errorf("Error attempting to close timed out connection: %s: %+v", request.GetConnection().GetId(), err)
+
+	conn = conn.Clone()
+	timerCtx := extend.WithValuesFromContext(context.Background(), t.ctx)
+
+	timer := &timer{
+		stopCh: make(chan struct{}),
+	}
+	timer.timer = time.AfterFunc(time.Until(expireTime), func() {
+		executor, ok := t.executors.Load(conn.GetId())
+		if !ok {
+			return
 		}
-	}), nil
+
+		executor.AsyncExec(func() {
+			select {
+			case <-timer.stopCh:
+				logEntry.Warnf("timer has been already stopped: %v", conn.GetId())
+			default:
+				if err := t.close(timerCtx, conn, next.Server(ctx)); err != nil {
```

Comment:
What if the timer has been stopped here? I'm not sure that timer.stopCh is a good solution for this case.

Reply:
It is a guard for such case:
```
1. -> request        : locking executor
2. -> timeout close  : waiting on executor
3. -request->        : stopping timer (oops! it is too late), closing stopCh
4. request ->        : unlocking executor
5. -timeout close->  : locking executor, checking stopCh - it is closed, returning without actual Close
```